src/mount-support.c

```diff
@@ -276,14 +277,47 @@ void setup_snappy_os_mounts()
 	// moved all paths to /tmp/snap.rootfs_*. Because we are using unshare with
 	// CLONE_NEWNS we can essentially use pivot_root just like chroot but this
 	// makes apparmor unaware of the old root so everything works okay.
+
+	// The code below was taken fomr src/lxc/conf.c in github.com/lxc/lxc
+	// It is available under the terms of the LGPLv2.1 and higher and
+	// was written by various LXC developers.
+
 	debug("chrooting into %s", rootfs_dir);
-	if (chdir(rootfs_dir) == -1) {
-		die("cannot change working directory to %s", rootfs_dir);
+
+	int oldroot = -1, newroot = -1;
+	oldroot = open("/", O_DIRECTORY | O_RDONLY);
+	if (oldroot < 0) {
+		die("failed to open the old root filesystem");
+	}
+
+	newroot = open(rootfs_dir, O_DIRECTORY | O_RDONLY);
```

Comment:
The problem with O_PATH is that it didn't exist at all until 2.6.39 and didn't support close() or fchdir() which I use below until the 3.5 kernel.

If snap-confine isn't expected to run with a kernel below 3.5 (say random Android kernels which were based on 3.1 or 3.4), then O_PATH would indeed be fine.